plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestHiveConnectorTest.java

```diff
@@ -3115,6 +3115,62 @@ public void testPartitionPerScanLimit()
         assertUpdate("DROP TABLE " + tableName);
     }
 
+    @Test
+    public void testPartitionPerScanLimitWithMultiplePartitionColumns()
+    {
+        String tableName = "test_multi_partition_per_scan_limit";
+        String partitionsTable = "\"" + tableName + "$partitions\"";
+
+        assertUpdate("" +
+                "CREATE TABLE " + tableName + " (  foo varchar,   part1 bigint,   part2 bigint) " +
+                "WITH (partitioned_by = ARRAY['part1', 'part2'])");
+
+        // insert 1200 partitions with part1 being NULL
+        for (int batchNumber = 0; batchNumber < 12; batchNumber++) {
+            // insert with a loop to avoid max open writers limit
+            assertUpdate(
+                    format(
+                            "INSERT INTO %s SELECT 'bar' foo, NULL part1, n part2 FROM UNNEST(sequence(%s, %s)) a(n)",
+                            tableName,
+                            batchNumber * 100 + 1,
+                            (batchNumber + 1) * 100),
+                    100);
+        }
+        // insert 10 partitions with part1=part2
+        assertUpdate("INSERT INTO " + tableName + " SELECT 'bar' foo, n part1, n part2 FROM UNNEST(sequence(1, 10)) a(n)", 10);
+
+        // verify can query 1000 partitions
+        assertThat(query("SELECT count(*) FROM " + tableName + " WHERE part1 IS NULL AND part2 < 1001"))
+                .matches("VALUES BIGINT '1000'");
+
+        // verify cannot query more than 1000 partitions
+        assertThatThrownBy(() -> query("SELECT count(*) FROM " + tableName + " WHERE part1 IS NULL AND part2 <= 1001"))
+                .hasMessage(format("Query over table 'tpch.%s' can potentially read more than 1000 partitions", tableName));
+        assertThatThrownBy(() -> query("SELECT count(*) FROM " + tableName))
+                .hasMessage(format("Query over table 'tpch.%s' can potentially read more than 1000 partitions", tableName));
+
+        // verify we can query with a predicate that is not representable as a TupleDomain
+        // TODO this shouldn't fail
+        assertThatThrownBy(() -> query("SELECT * FROM " + tableName + " WHERE part1 % 400 = 3")) // may be translated to Domain.all
```

Comment:
`may be translated to Domain.all`; why would it allow nulls?

Reply:
Can we raise an issue that tracks the changes like if a function like this is applied them we could create a `Domain` like `Domain.all except nulls` not sure if it is feasible.

Reply:
> may be translated to Domain.all; why would it allow nulls?

@losipiuk correct, we *could* infer non-null, although i believe we don't do that today.

actually, it would be simple to do, for any expression with one free variable.